### PR TITLE
Fix nil pointer panic in StatefulSet status reporting during informer cache races

### PR DESCRIPTION
**Pull request by WHOIM1205**
## Summary

This PR fixes a **nil pointer dereference panic** in the Prometheus Operator that can occur during **informer cache inconsistencies**, especially during operator startup, restarts, or namespace deletion.

The root cause was an invalid contract in `informers.ForResource.Get()` which could return `(nil, nil)`. Callers assumed a non-nil object when `err == nil`, leading to a panic when dereferencing the returned object.

This PR fixes the root cause and adds defensive checks to prevent similar crashes in the future.

---

## Problem

While reporting Prometheus status, the operator retrieves StatefulSets from the informer cache. Under real Kubernetes race conditions (startup before cache sync, informer reinitialization, namespace deletion), the informer `Get()` method may return `(nil, nil)`.

This leads to a nil pointer dereference when the returned object is accessed, crashing the operator.

---

## Root Cause

- `informers.ForResource.Get()` could return `(nil, nil)` when:
  - No informers were configured
  - Informer cache had not yet synced
- Callers incorrectly assumed:
  - `err == nil` implies a valid object
- This assumption breaks under normal Kubernetes lifecycle races

---

### Root Cause Fix
- `ForResource.Get()` now returns a proper `NotFound` error when no informers are available or no object is found.
- This restores a safe and predictable contract for all callers.

### Defense in Depth
- Added defensive `nil` checks when retrieving StatefulSets in:
  - Prometheus operator
  - Alertmanager operator
  - Thanos Ruler operator
- Nil objects are now handled as `NotFound` instead of causing a panic.

---

## How to Reproduce

1. Install Prometheus Operator with namespace restrictions:
   ```bash
   helm install prometheus-operator prometheus-community/kube-prometheus-stack \
     --set prometheusOperator.namespaces.allowList="{monitoring,app-team-a}"
   ```

2. Create a Prometheus resource:
   ```yaml
   apiVersion: monitoring.coreos.com/v1
   kind: Prometheus
   metadata:
     name: test
     namespace: monitoring
   spec:
     replicas: 1
   ```

3. During operator startup (before informer cache sync completes), delete the namespace:
   ```bash
   kubectl delete namespace monitoring &
   ```

4. Observe operator behavior:
   - **Before fix:** operator panics due to nil pointer dereference
   - **After fix:** operator continues running and handles the case gracefully

---

### Before
- Operator crash and restart
- Temporary monitoring reconciliation outage
- Status reporting fails unexpectedly

### After
- No panics during informer cache races
- Stable operator during startup, restarts, and deletions
- Graceful degradation instead of crash

---

## Why This Is Safe

- No reconciliation logic changes
- Normal code paths are unaffected
- Fix only applies to previously unsafe edge cases
- Low risk and safe to backport

i have also added some test cases  : 
Expected output:
=== RUN   TestInformers
=== RUN   TestInformers/TestGet
=== RUN   TestInformers/TestGetWithEmptyInformers
=== RUN   TestInformers/TestGetNeverReturnsNilNil
=== RUN   TestInformers/TestGetNeverReturnsNilNil/empty_informers
=== RUN   TestInformers/TestGetNeverReturnsNilNil/object_not_found_in_single_informer
=== RUN   TestInformers/TestGetNeverReturnsNilNil/object_not_found_in_multiple_informers
=== RUN   TestInformers/TestGetReturnsObjectWhenFound
=== RUN   TestInformers/TestGetSearchesAllInformers
--- PASS: TestInformers (0.00s)
    --- PASS: TestInformers/TestGet (0.00s)
    --- PASS: TestInformers/TestGetWithEmptyInformers (0.00s)
    --- PASS: TestInformers/TestGetNeverReturnsNilNil (0.00s)
    --- PASS: TestInformers/TestGetReturnsObjectWhenFound (0.00s)
    --- PASS: TestInformers/TestGetSearchesAllInformers (0.00s)
PASS

### CODE REVIEW

**Comment by WHOIM1205** [comment]
hi @simonpasquier 

This PR fixes a real nil pointer panic caused by informer cache races.

`informers.ForResource.Get()` could return `(nil, nil)` during startup, cache resync, or namespace deletion. Callers assumed a non-nil object on a nil error, leading to a panic when dereferencing StatefulSets in status reporting.

The fix restores a safe contract by returning a proper `NotFound` error and adds defensive nil checks in Prometheus, Alertmanager, and Thanos paths to prevent crashes.

This preserves existing behavior while eliminating a production crash scenario.

File: pkg/informers/informers.go [hunk 1]
```diff
@@ -189,8 +189,16 @@ func (w *ForResource) ListAllByNamespace(namespace string, selector labels.Selec
 }
 
 // Get invokes all wrapped informers and returns the first found runtime object.
-// It returns the first occurred error.
+// It returns a NotFound error if the object isn't found in any informer or if
+// there are no informers configured.
 func (w *ForResource) Get(name string) (runtime.Object, error) {
+	// Return a proper NotFound error if no informers are configured.
```
Comment (simonpasquier) [inline pkg/informers/informers.go]: no informers being configured is a programming error which should trigger a panic.

File: pkg/alertmanager/operator.go [hunk 2]
```diff
@@ -708,6 +708,13 @@ func (c *Operator) getStatefulSetFromAlertmanagerKey(key string) (*appsv1.Statef
 		return nil, fmt.Errorf("failed to retrieve StatefulSet from informer: %w", err)
 	}
 
+	// Defensive nil check: if informer returns nil without error
+	// (e.g., during cache inconsistency), treat as not found.
+	if obj == nil {
```
Comment (simonpasquier) [inline pkg/alertmanager/operator.go]: With the change to informers.go, is it now possible for Get() to return a nil value?

File: pkg/informers/informers.go (hunk 1 above)
Comment (WHOIM1205) [reply pkg/informers/informers.go]: thanks for the clarification @simonpasquier 
I’ve updated `ForResource.Get()` to panic when it is called with no informers configured, as this is indeed a programming error and should fail fast rather than being handled as NotFound.
The rest of the logic remains unchanged. Please let me know if you’d like me to further simplify the defensive checks now that the contract is stricter.

File: pkg/alertmanager/operator.go (hunk 2 above)
Comment (WHOIM1205) [reply pkg/alertmanager/operator.go]: With the proposed change to panic on empty informers, Get() should no longer return (nil, nil) in normal operation.
The defensive nil checks were added to guard against unexpected informer behavior and future regressions, but I’m happy to remove or scope them if you’d prefer relying solely on the stricter Get() contract.

File: pkg/informers/informers.go [hunk 3]
```diff
@@ -189,8 +189,13 @@ func (w *ForResource) ListAllByNamespace(namespace string, selector labels.Selec
 }
 
 // Get invokes all wrapped informers and returns the first found runtime object.
-// It returns the first occurred error.
+// It returns a NotFound error if the object isn't found in any informer.
+// It panics if no informers are configured, as this is a programming error.
 func (w *ForResource) Get(name string) (runtime.Object, error) {
+	if len(w.informers) == 0 {
```
Comment (simonpasquier) [inline pkg/informers/informers.go]: sorry I meant to say that when the list of informers is empty, we should return a notfound error.

File: pkg/informers/informers.go [hunk 4]
```diff
@@ -206,6 +211,12 @@ func (w *ForResource) Get(name string) (runtime.Object, error) {
 		return ret, nil
 	}
 
+	// Ensure we always return a NotFound error if the object wasn't found,
+	// even if err happens to be nil (which shouldn't happen, but be defensive).
+	if err == nil {
+		return nil, apierrors.NewNotFound(schema.GroupResource{}, name)
+	}
+
 	return nil, err
```
Comment (simonpasquier) [inline pkg/informers/informers.go]: when we reach this part, we know that the object wasn't found. We also need to keep track of the resource being watched to avoid the empty schema.GroupResource{}.
```suggestion
	return nil, apierrors.NewNotFound(schema.GroupResource{}, name)
```

File: pkg/alertmanager/operator.go (hunk 2 above)
Comment (simonpasquier) [reply pkg/alertmanager/operator.go]: The new contract of `*ForResource.Get()` is that it never returns a nil pointer. A regression would be easier to detect if we don't do defensive programming.

Comment (WHOIM1205) [reply pkg/alertmanager/operator.go]: Thanks for the clarification.
I’ve removed the defensive nil checks and updated the code to fully rely on the `ForResource.Get()` contract that it never returns a nil object. This should make any regression fail fast and be easier to detect.
Please let me know if there’s anything else you’d like adjusted.

File: pkg/informers/informers.go (hunk 3 above)
Comment (WHOIM1205) [reply pkg/informers/informers.go]: Thanks for the clarification — I’ve updated `ForResource.Get()` to return a NotFound error when the informers list is empty and removed the panic. Please let me know if this matches the expected behavior now.

File: pkg/informers/informers.go [hunk 5]
```diff
@@ -189,8 +189,12 @@ func (w *ForResource) ListAllByNamespace(namespace string, selector labels.Selec
...
 
 // Get invokes all wrapped informers and returns the first found runtime object.
-// It returns the first occurred error.
+// It returns a NotFound error if the object isn't found in any informer.
 func (w *ForResource) Get(name string) (runtime.Object, error) {
+	if len(w.informers) == 0 {
+		return nil, apierrors.NewNotFound(schema.GroupResource{}, name)
+	}
+
 	var err error
```
Comment (simonpasquier) [inline pkg/informers/informers.go]: reading the code again, we don't need to declare `err` here. If we loop over all the informers and couldn't find the key, we should just return a NotFound error.

File: pkg/informers/informers.go [hunk 6]
```diff
@@ -206,6 +210,12 @@ func (w *ForResource) Get(name string) (runtime.Object, error) {
 		return ret, nil
 	}
 
+	// Ensure we always return a NotFound error if the object wasn't found,
+	// even if err happens to be nil (which shouldn't happen, but be defensive).
+	if err == nil {
+		return nil, apierrors.NewNotFound(schema.GroupResource{}, name)
```
Comment (simonpasquier) [inline pkg/informers/informers.go]: for consistency we should return the proper GroupResource.

File: pkg/informers/informers.go (hunk 5 above)
Comment (WHOIM1205) [reply pkg/informers/informers.go]: Good catch, thanks.

I’ve removed the unused `err` variable and simplified the logic so that after iterating over all informers, we directly return a NotFound error when the key isn’t found.

File: pkg/informers/informers.go (hunk 6 above)
Comment (WHOIM1205) [reply pkg/informers/informers.go]: Thanks — I’ve updated the NotFound return to use the proper GroupResource for consistency.

File: pkg/informers/informers.go [hunk 7]
```diff
@@ -189,13 +191,14 @@ func (w *ForResource) ListAllByNamespace(namespace string, selector labels.Selec
 }
 
 // Get invokes all wrapped informers and returns the first found runtime object.
-// It returns the first occurred error.
+// It returns a NotFound error if the object isn't found in any informer.
```
Comment (simonpasquier) [inline pkg/informers/informers.go:194]:
```suggestion
// It returns the first occurring error or a NotFound error if the object isn't found in any informer.
```

File: pkg/informers/informers.go [hunk 8]
```diff
@@ -189,13 +191,14 @@ func (w *ForResource) ListAllByNamespace(namespace string, selector labels.Selec
...
 func (w *ForResource) Get(name string) (runtime.Object, error) {
-	var err error
+	if len(w.informers) == 0 {
+		return nil, apierrors.NewNotFound(w.gr, name)
+	}
```
Comment (simonpasquier) [inline pkg/informers/informers.go]: not needed anymore
```suggestion
```

File: pkg/informers/informers.go [hunk 9]
```diff
@@ -206,7 +209,7 @@ func (w *ForResource) Get(name string) (runtime.Object, error) {
 		return ret, nil
```
Comment (simonpasquier) [inline pkg/informers/informers.go:205]:
```suggestion
		return ret, err
```

File: pkg/informers/informers.go
Comment (simonpasquier) [inline pkg/informers/informers.go:201]: we can directly  `return ret, err` here

Comment (WHOIM1205) [reply pkg/informers/informers.go:201]: Good point — I’ve simplified the code to directly return `ret, err` there.

File: pkg/informers/informers.go [hunk 10]
```diff
@@ -189,13 +191,14 @@ func (w *ForResource) ListAllByNamespace(namespace string, selector labels.Selec
 }
 
 // Get invokes all wrapped informers and returns the first found runtime object.
-// It returns the first occurred error.
+// It returns a NotFound error if the object isn't found in any informer.
 func (w *ForResource) Get(name string) (runtime.Object, error) {
-	var err error
+	if len(w.informers) == 0 {
+		return nil, apierrors.NewNotFound(w.gr, name)
+	}
```
Comment (simonpasquier) [inline pkg/informers/informers.go]:
```suggestion
```
